### ISSUE TITLE
Unhandled error in spell checking

### ISSUE THREAD

**Issue by paulkater** [vkbo/novelWriter#1096]
On starting NovelWriter, suddenly this popped up:

Environment:
novelWriter Version: 1.6.2
Host OS: linux (5.15.0-46-generic)
Python: 3.10.4 (0x30a04f0)
Qt: 5.15.3, PyQt: 5.15.6
lxml: 4.8.0
enchant: 3.2.2

Error:
Dict unusable: the underlying C-library object has been freed.

Traceback:
  File "/home/paul/apps/novelwriter/novelWriter.py", line 21, in <module>
    novelwriter.main(sys.argv[1:])

  File "/home/paul/apps/novelwriter/novelwriter/__init__.py", line 293, in main
    nwGUI.showProjectLoadDialog()

  File "/home/paul/apps/novelwriter/novelwriter/guimain.py", line 959, in showProjectLoadDialog
    self.openProject(dlgProj.openPath)

  File "/home/paul/apps/novelwriter/novelwriter/guimain.py", line 533, in openProject
    self.docEditor.setDictionaries()

  File "/home/paul/apps/novelwriter/novelwriter/gui/doceditor.py", line 694, in setDictionaries
    self.spEnchant.setLanguage(theLang, self.theProject.projDict)

  File "/home/paul/apps/novelwriter/novelwriter/core/spellcheck.py", line 78, in setLanguage
    self._theDict.add_to_session(pWord)

  File "/home/paul/.local/lib/python3.10/site-packages/enchant/__init__.py", line 670, in add_to_session
    self._check_this()

  File "/home/paul/.local/lib/python3.10/site-packages/enchant/__init__.py", line 594, in _check_this
    super()._check_this(msg)

  File "/home/paul/.local/lib/python3.10/site-packages/enchant/__init__.py", line 152, in _check_this
    raise Error(msg)

**Comment by paulkater** [comment]
This only happens with 1 writing project on my laptop. The others open just fine.
This project is synced with my desktop, on which it opens fine as well. Opening and closing it there, and waiting for the sync to the laptop (via NextCloud) did not fix the problem.

**Comment by paulkater** [comment]
There is at least one more project that suffers from this as well. Could it be that Nextcloud is corrupting something? Before switching to that I used to sync through SyncThing and this problem didn't appear there. Hmm...

**Comment by vkbo** [comment]
Hmm, the error seems to originate in the enchant package, which is the tool used for spell checking. Perhaps there is a corrupt dictionary file? Do the problematic projects have a language or spell checking setting that the others don't have?

In any case, I should wrap error handling on the call to the external package so that this error at least is handled.

**Comment by vkbo** [comment]
Sorry, I just responded to the first message. You have already identified what I asked I see. All I can do is add some code to handle the error better in novelWriter, which I will do. Perhaps there is a bug report here that can be filed with enchant as well? I'll investigate.

**Comment by paulkater** [comment]
Wrapping the error would be great. 

Yes, I'll see if I can find a place to report this issue with Enchant. Thanks for the swift response and great software.

**Comment by paulkater** [comment]
FYI: I reported a bug for PyEnchant about this.

**Comment by vkbo** [comment]
There may still be an error on my side of the code, or an invalid state due to some content of your custom dictionary, so let's rule out that first.

**Comment by vkbo** [comment]
It seems to be crashing when the "meta/wordlist.txt" file is loaded from your project folder. Is there anything wrong with the content of that file? Does the other project have one (that isn't empty)? These words are loaded immediately after the dictionary is set up, and I delete the old one before that. Maybe there is an issue related to that.

**Comment by vkbo** [comment]
Since I cannot recreate the error, could you test a thing for me to debug this?

Go into your file `/home/paul/apps/novelwriter/novelwriter/core/spellcheck.py`
Comment out the three lines around line 60 like I've done in this screenshot:
![image](https://user-images.githubusercontent.com/1619840/185189279-cb34b1c8-f09d-4e05-b2f3-fdc41e1caee6.png)

The try again, and see if it work.

**Comment by paulkater** [comment]
1. Installed pyenchant again and confirmed that the problem occurs again with the 1st project.
2. Commented the 3 lines and tried again. Problem persists.
3. meta/wordlist.txt for the 1st project contains 2 words
4. meta/wordlist.txt for the 2nd project contains 1 word
5. Removed the 1 word from wordlist.txt of the 2nd project and that then opens without a problem.
6. Added the 1 word back and things go belly up again.

**Comment by paulkater** [comment]
Enabled the 3 statements, removed the 1 word, no problem.
Sorry, should have tested that too before commenting.
I'm going to check if there's a diff between the version on the desktop and the laptop, but I'm quite confident they're both the latest.

**Comment by paulkater** [comment]
Pulled the latest again and installed that fresh on laptop. Problem persists.
Both machines also run python 3.10.4 (to rule that out)

**Comment by vkbo** [comment]
I'm running pyenchant 3.2.0. I'll try the latest pip version and see if I can reproduce this. There is a change to the code freeing up resources in their release 3.2.1, so it's worth trying.

**Comment by paulkater** [comment]
Their latest is 3.2.2. which is on the laptop. I put that on the desktop too (that had 3.2.0).
No problem on the desktop version of novelWriter (which is sooo weird).

**Comment by paulkater** [comment]
Did some experimenting:
spellcheck.py:

        self._readProjectDictionary(projectDict)
        for pWord in self._projDict:
           try:
              self._theDict.add_to_session(pWord)
           except:
              print ("fail")
        return


![image](https://user-images.githubusercontent.com/111410394/185197732-17ebfd38-b269-4cc4-ba0f-6714d314cccc.png)

Does this help any?
(This is from the project with 2 words in the wordlist.)

**Comment by vkbo** [comment]
Hmm, the issue may be somewhere else entirely. Could you run novelWriter.py again with the --debug flag? I would like to see where the other two errors come from.

**Comment by paulkater** [comment]
~/apps/novelwriter$ ./novelWriter.py --debug
[2022-08-17 19:44:00,245]        __init__.py:207   INFO      Starting novelWriter 1.6.2 (0x010602f0) 2022-03-20
[2022-08-17 19:44:00,245]          config.py:260   DEBUG     Initialising Config ...
[2022-08-17 19:44:00,245]          config.py:312   DEBUG     Assets: /home/paul/apps/novelwriter/novelwriter/assets
[2022-08-17 19:44:00,245]          config.py:430   DEBUG     Loading config file
[2022-08-17 19:44:00,257]          config.py:982   DEBUG     Checking package 'pyenchant': OK
[2022-08-17 19:44:00,257]          config.py:369   DEBUG     Found manual: /home/paul/apps/novelwriter/novelwriter/assets/manual.pdf
[2022-08-17 19:44:00,257]          config.py:372   DEBUG     Config initialisation complete
[2022-08-17 19:44:00,330]          config.py:394   DEBUG     Loaded: /usr/share/qt5/translations/qtbase_en
[2022-08-17 19:44:00,330]          config.py:394   DEBUG     Loaded: /usr/share/qt5/translations/qtbase_en_US
[2022-08-17 19:44:00,330]          config.py:394   DEBUG     Loaded: /usr/share/qt5/translations/qtbase_en_Latn_US
[2022-08-17 19:44:00,331]          config.py:394   DEBUG     Loaded: /home/paul/apps/novelwriter/novelwriter/assets/i18n/nw_en_US
[2022-08-17 19:44:00,331]         guimain.py:66    DEBUG     Initialising GUI ...
[2022-08-17 19:44:00,331]         guimain.py:74    INFO      OS: linux
[2022-08-17 19:44:00,331]         guimain.py:75    INFO      Kernel: 5.15.0-46-generic
[2022-08-17 19:44:00,331]         guimain.py:76    INFO      Host: Aspiratie
[2022-08-17 19:44:00,331]         guimain.py:77    INFO      Qt5: 5.15.3 (51503)
[2022-08-17 19:44:00,331]         guimain.py:78    INFO      PyQt5: 5.15.6 (51506)
[2022-08-17 19:44:00,331]         guimain.py:79    INFO      Python: 3.10.4 (0x30a04f0)
[2022-08-17 19:44:00,331]         guimain.py:80    INFO      GUI Language: en_US
[2022-08-17 19:44:00,341]           theme.py:245   INFO      Loading GUI theme 'default'
[2022-08-17 19:44:00,341]           theme.py:306   INFO      Loading syntax theme 'default_light'
[2022-08-17 19:44:00,341]           theme.py:526   INFO      Loading icon theme 'typicons_light'
[2022-08-17 19:44:00,342]           theme.py:564   DEBUG     Scanning theme icons
[2022-08-17 19:44:00,344]       statusbar.py:47    DEBUG     Initialising GuiMainStatus ...
[2022-08-17 19:44:00,349]       statusbar.py:118   DEBUG     GuiMainStatus initialisation complete
[2022-08-17 19:44:00,349]        projtree.py:59    DEBUG     Initialising GuiProjectTree ...
[2022-08-17 19:44:00,352]        projtree.py:131   DEBUG     GuiProjectTree initialisation complete
[2022-08-17 19:44:00,352]       noveltree.py:49    DEBUG     Initialising GuiNovelTree ...
[2022-08-17 19:44:00,352]       noveltree.py:100   DEBUG     GuiNovelTree initialisation complete
[2022-08-17 19:44:00,352]       doceditor.py:76    DEBUG     Initialising GuiDocEditor ...
[2022-08-17 19:44:00,353]       doceditor.py:2567  DEBUG     Initialising GuiDocEditHeader ...
[2022-08-17 19:44:00,353]       doceditor.py:2667  DEBUG     GuiDocEditHeader initialisation complete
[2022-08-17 19:44:00,353]       doceditor.py:2791  DEBUG     Initialising GuiDocEditFooter ...
[2022-08-17 19:44:00,354]       doceditor.py:2889  DEBUG     GuiDocEditFooter initialisation complete
[2022-08-17 19:44:00,354]       doceditor.py:2179  DEBUG     Initialising GuiDocEditSearch ...
[2022-08-17 19:44:00,359]       doceditor.py:2349  DEBUG     GuiDocEditSearch initialisation complete
[2022-08-17 19:44:00,359]      spellcheck.py:44    DEBUG     Enchant spell checking activated
[2022-08-17 19:44:00,359]    dochighlight.py:52    DEBUG     Initialising GuiDocHighlighter ...
[2022-08-17 19:44:00,360]    dochighlight.py:88    DEBUG     Setting up highlighting rules
[2022-08-17 19:44:00,360]    dochighlight.py:80    DEBUG     GuiDocHighlighter initialisation complete
[2022-08-17 19:44:00,360]      spellcheck.py:69    DEBUG     Enchant spell checking for language '' loaded
[2022-08-17 19:44:00,360]      spellcheck.py:147   ERROR     Failed to extract information about the dictionary
[2022-08-17 19:44:00,361]           error.py:47    ERROR     AttributeError: 'NoneType' object has no attribute 'name'
[2022-08-17 19:44:00,365]    dochighlight.py:88    DEBUG     Setting up highlighting rules
[2022-08-17 19:44:00,365]       doceditor.py:179   DEBUG     GuiDocEditor initialisation complete
[2022-08-17 19:44:00,365]       docviewer.py:1160  DEBUG     Initialising GuiDocViewDetails ...
[2022-08-17 19:44:00,366]       docviewer.py:1189  DEBUG     GuiDocViewDetails initialisation complete
[2022-08-17 19:44:00,366]       docviewer.py:55    DEBUG     Initialising GuiDocViewer ...
[2022-08-17 19:44:00,366]       docviewer.py:727   DEBUG     Initialising GuiDocViewHeader ...
[2022-08-17 19:44:00,366]       docviewer.py:828   DEBUG     GuiDocViewHeader initialisation complete
[2022-08-17 19:44:00,367]       docviewer.py:934   DEBUG     Initialising GuiDocViewFooter ...
[2022-08-17 19:44:00,368]       docviewer.py:1088  DEBUG     GuiDocViewFooter initialisation complete
[2022-08-17 19:44:00,368]       docviewer.py:87    DEBUG     GuiDocViewer initialisation complete
[2022-08-17 19:44:00,368]     itemdetails.py:44    DEBUG     Initialising GuiItemDetails ...
[2022-08-17 19:44:00,369]     itemdetails.py:188   DEBUG     GuiItemDetails initialisation complete
[2022-08-17 19:44:00,369]         outline.py:88    DEBUG     Initialising GuiOutline ...
[2022-08-17 19:44:00,369]         outline.py:127   DEBUG     GuiOutline initialisation complete
[2022-08-17 19:44:00,369]  outlinedetails.py:52    DEBUG     Initialising GuiOutlineDetails ...
[2022-08-17 19:44:00,371]  outlinedetails.py:236   DEBUG     GuiOutlineDetails initialisation complete
[2022-08-17 19:44:00,372]        mainmenu.py:47    DEBUG     Initialising GuiMainMenu ...
[2022-08-17 19:44:00,374]        mainmenu.py:70    DEBUG     GuiMainMenu initialisation complete
[2022-08-17 19:44:00,386]        projtree.py:679   DEBUG     Building the project tree ...
[2022-08-17 19:44:00,386]        projtree.py:687   DEBUG     0 item(s) added to the project tree
[2022-08-17 19:44:00,431]         guimain.py:297   DEBUG     GUI initialisation complete
[2022-08-17 19:44:00,431]         guimain.py:311   INFO      novelWriter is ready ...
[2022-08-17 19:44:00,431]        projload.py:59    DEBUG     Initialising GuiProjectLoad ...
[2022-08-17 19:44:00,436]        projload.py:150   DEBUG     GuiProjectLoad initialisation complete
[2022-08-17 19:44:04,210]         project.py:382   INFO      Opening project: /home/paul/Nextcloud/NovelwriterProjecten/Mirror Earth 5  Pathways
[2022-08-17 19:44:04,212]         project.py:528   DEBUG     Found project meta
[2022-08-17 19:44:04,212]         project.py:549   DEBUG     Found project settings
[2022-08-17 19:44:04,212]         project.py:590   DEBUG     Found project content
[2022-08-17 19:44:04,213]         options.py:84    DEBUG     Loading GUI options file
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 539, resource id: 12591111, major code: 40 (TranslateCoords), minor code: 0
[2022-08-17 19:44:04,215]         project.py:1355  DEBUG     Checking files in project content folder
[2022-08-17 19:44:04,215]         project.py:1373  DEBUG     Checking file 8aaf7e781c5fc.nwd, handle '8aaf7e781c5fc': OK
[2022-08-17 19:44:04,215]         project.py:1373  DEBUG     Checking file 1f15634472e9a.nwd, handle '1f15634472e9a': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file ed13f82a1e8ef.nwd, handle 'ed13f82a1e8ef': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 1c133c0d24ae0.nwd, handle '1c133c0d24ae0': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 853df946df410.nwd, handle '853df946df410': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file a6d4de9f25be0.nwd, handle 'a6d4de9f25be0': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 0cc9d5f279ab4.nwd, handle '0cc9d5f279ab4': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 552fa08c8d0ff.nwd, handle '552fa08c8d0ff': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 10138848063eb.nwd, handle '10138848063eb': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 982720adde3c4.nwd, handle '982720adde3c4': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file eaede8e9e1452.nwd, handle 'eaede8e9e1452': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file a43293ecdf5ac.nwd, handle 'a43293ecdf5ac': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 8453cefa555ba.nwd, handle '8453cefa555ba': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 4e40ce9401ffe.nwd, handle '4e40ce9401ffe': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 53360658a81ac.nwd, handle '53360658a81ac': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 8979136e6ffb6.nwd, handle '8979136e6ffb6': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 5528b95e8377d.nwd, handle '5528b95e8377d': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file cf85c725f416b.nwd, handle 'cf85c725f416b': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file a364fece04c4c.nwd, handle 'a364fece04c4c': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 7b6260462ebb7.nwd, handle '7b6260462ebb7': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file ed45d906020d5.nwd, handle 'ed45d906020d5': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 0b2b669c78de5.nwd, handle '0b2b669c78de5': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 57942d2bd999b.nwd, handle '57942d2bd999b': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 42aeeda08f25d.nwd, handle '42aeeda08f25d': OK
[2022-08-17 19:44:04,216]         project.py:1373  DEBUG     Checking file 18fc8b04e2fb3.nwd, handle '18fc8b04e2fb3': OK
[2022-08-17 19:44:04,216]         project.py:1384  DEBUG     File check OK
[2022-08-17 19:44:04,216]           index.py:146   DEBUG     Loading index file
[2022-08-17 19:44:04,217]           index.py:669   DEBUG     Checking index
[2022-08-17 19:44:04,217]        projtree.py:679   DEBUG     Building the project tree ...
[2022-08-17 19:44:04,222]        projtree.py:687   DEBUG     31 item(s) added to the project tree
[2022-08-17 19:44:04,222]        projtree.py:379   DEBUG     Saving project tree item order
[2022-08-17 19:44:04,223]      spellcheck.py:63    DEBUG     Deleting old pyenchant broker
[2022-08-17 19:44:04,223]      spellcheck.py:69    DEBUG     Enchant spell checking for language '' loaded
[2022-08-17 19:44:04,223]      spellcheck.py:172   DEBUG     Loading project word list
[2022-08-17 19:44:04,223]      spellcheck.py:178   DEBUG     Project word list contains 2 words
fail
fail
[2022-08-17 19:44:04,223]      spellcheck.py:147   ERROR     Failed to extract information about the dictionary
[2022-08-17 19:44:04,223]           error.py:47    ERROR     AttributeError: 'NoneType' object has no attribute 'name'
[2022-08-17 19:44:04,224]        document.py:85    DEBUG     Opening document: 7b6260462ebb7.nwd
[2022-08-17 19:44:04,249]       doceditor.py:370   DEBUG     Document highlighted in 20.550 ms
[2022-08-17 19:44:04,263]         guimain.py:558   DEBUG     Project load complete

**Comment by vkbo** [comment]
Well, there's your problem!

`[2022-08-17 19:44:04,223] spellcheck.py:69 DEBUG Enchant spell checking for language '' loaded`

I forced language to be an empty string at my end, and I could reproduce the error. Surprisingly, enchant accepts that as a language input, but fails later on various other calls.

I'm not exactly sure what is causing the issue. Since it's isolated to one computer, it may be something wrong either with your spell check dictionaries, or something missing on your system. Could you go into Preferences and change language to another one, just to test?

**Comment by vkbo** [comment]
In any case, pyenchant recommends checking if a dictionary exists before loading it, so I will improve that section of the code so it is more robust. There should be an error popping up, but it shouldn't crash the app.

**Comment by paulkater** [comment]
Bingo.

[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file 7b6260462ebb7.nwd, handle '7b6260462ebb7': OK
[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file ed45d906020d5.nwd, handle 'ed45d906020d5': OK
[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file 0b2b669c78de5.nwd, handle '0b2b669c78de5': OK
[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file 57942d2bd999b.nwd, handle '57942d2bd999b': OK
[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file 42aeeda08f25d.nwd, handle '42aeeda08f25d': OK
[2022-08-17 19:57:09,464]         project.py:1373  DEBUG     Checking file 18fc8b04e2fb3.nwd, handle '18fc8b04e2fb3': OK
[2022-08-17 19:57:09,464]         project.py:1384  DEBUG     File check OK
[2022-08-17 19:57:09,464]           index.py:146   DEBUG     Loading index file
[2022-08-17 19:57:09,464]           index.py:669   DEBUG     Checking index
[2022-08-17 19:57:09,465]        projtree.py:679   DEBUG     Building the project tree ...
[2022-08-17 19:57:09,468]        projtree.py:687   DEBUG     31 item(s) added to the project tree
[2022-08-17 19:57:09,469]        projtree.py:379   DEBUG     Saving project tree item order
[2022-08-17 19:57:09,469]      spellcheck.py:63    DEBUG     Deleting old pyenchant broker
[2022-08-17 19:57:09,505]      spellcheck.py:69    DEBUG     Enchant spell checking for language 'en' loaded
[2022-08-17 19:57:09,510]      spellcheck.py:172   DEBUG     Loading project word list
[2022-08-17 19:57:09,510]      spellcheck.py:178   DEBUG     Project word list contains 2 words
[2022-08-17 19:57:09,511]        document.py:85    DEBUG     Opening document: 7b6260462ebb7.nwd
[2022-08-17 19:57:09,534]       doceditor.py:370   DEBUG     Document highlighted in 19.096 ms
[2022-08-17 19:57:09,550]         guimain.py:558   DEBUG     Project load complete

I changed the language from US English to British English and the problem is gone. (This is the same project with 2 words in wordlist).

Good sleuthing! I'll leave it on British (it's what I learnt in school too ;-).

As a test I added another word to wordlist (debug showed me doing it), reloaded the project and all is well.

Thank you for looking into this so quickly!

**Comment by vkbo** [comment]
Great! I'm still curious to know how the language setting ended up as an empty string. In any case, I will make a few fixes and add some tests to make sure these cases are handled.

**Comment by vkbo** [comment]
Update: If you clear the `spellcheck` value in the config file, this issue is triggered. That is the core problem I think.

**Comment by vkbo** [comment]
Fixed on the `patch` branch for release in 1.6.3.

**Comment by paulkater** [comment]
> I'm still curious to know how the language setting ended up as an empty string. 

I have no clue. The laptop previously ran the latest Fedora, a clean install, and I had no problems with that. I reverted to Ubuntu 22.04, another clean and straight-forward install, and that's when this happened. I guess there's some glitch in the Ubuntu installer that made this happen.

The desktop upgraded from Ubuntu 20.04 to 22.04, so all settings were retained, hence that had no problem.

*Bottom line*: Glad this got fixed, making novelWriter even better. :smile:
